File: src/webgl_data/glbuffer.fs
namespace Wil.Webgl.Data

open Fable.Core
open Wil.Js
open Wil.Webgl.Types
open Wil.Webgl.Core

type GlBuffer(gl: GL) =
  let mutable isBuffered = false
  let mutable dirtyOffset = System.Int32.MaxValue
  let mutable dirtyEndOffset = -1
  let mutable target = GlBufferTarget.ARRAY_BUFFER
  let mutable usage = GlBufferUsage.STATIC_DRAW
  let mutable factory = float32ArrayFactory
  let mutable data: JS.TypedArray = factory.Create(0)
  let mutable buffer = getEmptyBuffer gl

  member val BufferKind = target
  member val Usage = usage
  member _.Data with get() = data
  member val AutoClean = true with get, set

  member _.Init(bufferTarget, bufferUsage, arrayFactory) =
    target <- bufferTarget
    usage <- bufferUsage
    factory <- arrayFactory
    buffer <- gl.createBuffer()

  member _.Delete() =
    let empty = getEmptyBuffer gl

    if buffer <> empty then
      gl.deleteBuffer(buffer)

  member _.Clean(buffered) =
    dirtyOffset <- System.Int32.MaxValue
    dirtyEndOffset <- -1
    isBuffered <- buffered

  member _.DirtyRange(startOffset, endOffset) =
    dirtyOffset <- min startOffset dirtyOffset
    dirtyEndOffset <- max endOffset dirtyEndOffset
    dirtyOffset <- max dirtyOffset 0
    dirtyEndOffset <- min dirtyEndOffset (data.length - 1)

  member this.SetLength(length) =
    if length <> data.length then
      let length = max length 0
      data <- factory.Create(length)
      isBuffered <- false
      this.DirtyRange(0, length - 1)

  member _.Bind() = bindBuffer gl target buffer
  member _.BindBase(location) = bindBufferBase gl target location buffer

  member this.SetValuesRange<'a>(values: 'a [], startOffset, endOffset) =
    this.DirtyRange(startOffset, endOffset)
    let dataArray = data :?> JS.TypedArray<'a>

    for i in [startOffset .. endOffset] do
      dataArray.[i] <- values.[i]

  member this.SetValuesOffset<'a>(values: 'a [], startOffset) =
    this.DirtyRange(startOffset, startOffset + values.Length - 1)
    (data :?> JS.TypedArray<'a>).set(values, startOffset) |> ignore

  member this.SetValues<'a>(values: 'a [], startOffset, endOffset) =
    if data.length <> values.Length then
      data <- factory.Create(values)
      isBuffered <- false
    else
      (data :?> JS.TypedArray<'a>).set(values) |> ignore
    
    this.DirtyRange(startOffset, endOffset)

  member this.SetValues<'a>(values: 'a []) = this.SetValues(values, 0, values.Length - 1)

  member this.Update() =
    if dirtyEndOffset >= 0 then
      if isBuffered then
        let dirtyLength = dirtyEndOffset - dirtyOffset + 1
        let bytesPerIndex = data.byteLength / data.length
        let startOffset = dirtyOffset * bytesPerIndex
        // let d = JS.Constructors.Float32Array.Create(data.buffer)
        // printfn $"{d}"
        bufferSubData gl target startOffset data dirtyOffset dirtyLength
      else
        // let d = JS.Constructors.Float32Array.Create(data.buffer)
        // printfn $"{d}"
        isBuffered <- this.AutoClean
        bufferData gl target data usage

    if this.AutoClean then this.Clean(true)

File: JS/webgl_data/glbuffer.js
import { float32ArrayFactory } from "../js/typedarray_utils.js";
import { GlCommon_bufferData, GlCommon_bufferSubData, GlCommon_bindBufferBase, GlCommon_bindBuffer, GlCommon_getEmptyBuffer } from "../webgl_core/webgl_browser_types.js";
import { class_type } from "../.fable/fable-library.3.0.0/Reflection.js";
import { max, comparePrimitives, min, equals } from "../.fable/fable-library.3.0.0/Util.js";
import { rangeNumber, getEnumerator } from "../.fable/fable-library.3.0.0/Seq.js";
import { ofSeq } from "../.fable/fable-library.3.0.0/List.js";

export class GlBuffer {
    constructor(gl) {
        this.gl = gl;
        this.isBuffered = false;
        this.dirtyOffset = 2147483647;
        this.dirtyEndOffset = -1;
        this.target = 34962;
        this.usage = 35044;
        this.factory = float32ArrayFactory;
        this.data = this.factory.Create(0);
        this.buffer = GlCommon_getEmptyBuffer(this.gl);
        this["BufferKind@"] = this.target;
        this["Usage@"] = this.usage;
        this["AutoClean@"] = true;
    }
}

export function GlBuffer$reflection() {
    return class_type("Wil.Webgl.Data.GlBuffer", void 0, GlBuffer);
}

export function GlBuffer_$ctor_10C6D16C(gl) {
    return new GlBuffer(gl);
}

export function GlBuffer__get_BufferKind(__) {
    return __["BufferKind@"];
}

export function GlBuffer__get_Usage(__) {
    return __["Usage@"];
}

export function GlBuffer__get_Data(_) {
    return _.data;
}

export function GlBuffer__get_AutoClean(__) {
    return __["AutoClean@"];
}

export function GlBuffer__set_AutoClean_Z1FBCCD16(__, v) {
    __["AutoClean@"] = v;
}

export function GlBuffer__Init_71B7C75E(_, bufferTarget, bufferUsage, arrayFactory) {
    _.target = bufferTarget;
    _.usage = bufferUsage;
    _.factory = arrayFactory;
    _.buffer = _.gl.createBuffer();
}

export function GlBuffer__Delete(_) {
    const empty = GlCommon_getEmptyBuffer(_.gl);
    if (!equals(_.buffer, empty)) {
        _.gl.deleteBuffer(_.buffer);
    }
}

export function GlBuffer__Clean_Z1FBCCD16(_, buffered) {
    _.dirtyOffset = 2147483647;
    _.dirtyEndOffset = -1;
    _.isBuffered = buffered;
}

export function GlBuffer__DirtyRange_Z37302880(_, startOffset, endOffset) {
    _.dirtyOffset = min(comparePrimitives, startOffset, _.dirtyOffset);
    _.dirtyEndOffset = max(comparePrimitives, endOffset, _.dirtyEndOffset);
    _.dirtyOffset = max(comparePrimitives, _.dirtyOffset, 0);
    _.dirtyEndOffset = min(comparePrimitives, _.dirtyEndOffset, _.data.length - 1);
}

export function GlBuffer__SetLength_Z524259A4(this$, length) {
    if (length !== this$.data.length) {
        const length_1 = max(comparePrimitives, length, 0) | 0;
        this$.data = this$.factory.Create(length_1);
        this$.isBuffered = false;
        GlBuffer__DirtyRange_Z37302880(this$, 0, length_1 - 1);
    }
}

export function GlBuffer__Bind(_) {
    GlCommon_bindBuffer(_.gl, _.target, _.buffer);
}

export function GlBuffer__BindBase_Z524259A4(_, location) {
    GlCommon_bindBufferBase(_.gl, _.target, location, _.buffer);
}

export function GlBuffer__SetValuesRange_Z3EAA4E5D(this$, values, startOffset, endOffset) {
    GlBuffer__DirtyRange_Z37302880(this$, startOffset, endOffset);
    const dataArray = this$.data;
    const enumerator = getEnumerator(ofSeq(rangeNumber(startOffset, 1, endOffset)));
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const i = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]() | 0;
            dataArray[i]=values[i];
        }
    }
    finally {
        enumerator.Dispose();
    }
}

export function GlBuffer__SetValuesOffset_Z59CA6BE1(this$, values, startOffset) {
    GlBuffer__DirtyRange_Z37302880(this$, startOffset, (startOffset + values.length) - 1);
    const value = this$.data.set(values, startOffset);
    void value;
}

export function GlBuffer__SetValues_Z3EAA4E5D(this$, values, startOffset, endOffset) {
    if (this$.data.length !== values.length) {
        this$.data = this$.factory.Create(values);
        this$.isBuffered = false;
    }
    else {
        const value = this$.data.set(values);
        void value;
    }
    GlBuffer__DirtyRange_Z37302880(this$, startOffset, endOffset);
}

export function GlBuffer__SetValues_5975E3(this$, values) {
    GlBuffer__SetValues_Z3EAA4E5D(this$, values, 0, values.length - 1);
}

export function GlBuffer__Update(this$) {
    if (this$.dirtyEndOffset >= 0) {
        if (this$.isBuffered) {
            const dirtyLength = ((this$.dirtyEndOffset - this$.dirtyOffset) + 1) | 0;
            const bytesPerIndex = (~(~(this$.data.byteLength / this$.data.length))) | 0;
            const startOffset = (this$.dirtyOffset * bytesPerIndex) | 0;
            GlCommon_bufferSubData(this$.gl, this$.target, startOffset, this$.data, this$.dirtyOffset, dirtyLength);
        }
        else {
            this$.isBuffered = GlBuffer__get_AutoClean(this$);
            GlCommon_bufferData(this$.gl, this$.target, this$.data, this$.usage);
        }
    }
    if (GlBuffer__get_AutoClean(this$)) {
        GlBuffer__Clean_Z1FBCCD16(this$, true);
    }
}

